src/osd/ReplicatedPG.cc

```diff
@@ -1496,6 +1496,22 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     return;
   }
 
+  // discard due to cluster full transition?  (we discard any op that
+  // originates before the cluster or pool is marked full; the client
+  // will resend after the full flag is removed or if they expect the
+  // op to succeed despite being full).
+  // FIXME: we exclude mds writes for now.
+  if (!m->get_source().is_mds() &&
+      info.history.last_epoch_marked_full > m->get_map_epoch()) {
+    dout(10) << __func__ << " discarding op sent before full " << m << " "
+	     << *m << dendl;
+    return;
+  }
+  if (osd->check_failsafe_full()) {
+    osd->reply_op_error(op, -ENOSPC);
```

Comment:
How about this:
- we continue to discard anything sent before the full epoch
- ops sent *while* full require a flag indicating they should be attempted (and may ENOSPC)
- ops sent *while* full without that flag are dropped, with a warning in the osd log (nobody should send these!)
- failsafe -> drop the request (for now)
- later, we need to make the failsafe do something smarter than hang client writes...  maybe a simple thing is that if/when the usage drops below the threshold we mark ourselves down?